File: packages/typescript/src/modules/tokens.ts
/**
 * TokensModule: tokenisation commands and saved card management.
 */

import { mapReturnCodeToError } from '../errors/index.js';
import type { EcollectConfig } from '../config.js';
import { BASE_URLS } from '../config.js';
import type {
  PaymentInfoItem,
  TokenCommand,
  TokenCommandRequest,
  TokenCommandResponse,
  QueryTokenRequest,
  QueryTokenResponse,
} from '../types/api.js';
import type { CardData, SavedCard } from '../types/index.js';
import type { HttpClient } from '../utils/http.js';
import type { SessionModule } from './session.js';
import { validateCardNumber, validateExpirationDate } from '../utils/validators.js';

function info(code: number, desc: string, value: string | number | undefined): PaymentInfoItem | null {
  if (value === undefined || value === null || value === '') return null;
  return { AttributeCode: code, AttributeDesc: desc, AttributeValue: String(value) };
}

function buildCardTokenInfoArray(card: CardData): PaymentInfoItem[] {
  const items: Array<PaymentInfoItem | null> = [
    info(0, 'CardNumber', card.cardNumber),
    info(2, 'PaymentSystem', card.paymentSystem),
    info(3, 'SecureCode', card.secureCode),
    info(4, 'ExpirationDate', card.expirationDate),
    info(6, 'Usermail', card.email),
    info(7, 'MobileCountryCode', card.mobileCountryCode),
    info(8, 'MobileNumber', card.mobileNumber),
    info(9, 'FiCode', card.fiCode),
    info(17, 'CardHolderName', card.cardHolderName),
    info(18, 'CardHolderIdType', card.cardHolderIdType),
    info(19, 'CardHolderId', card.cardHolderId),
    info(21, 'CardIssueBank', card.cardIssueBank),
    info(22, 'AccountType', card.accountType !== undefined ? String(card.accountType) : undefined),
    info(20, 'CardIssueCountry', card.cardIssueCountry),
  ];

  // CustomerId replaces cardholder details if present
  if (card.customerId) {
    items.push(info(100, 'CustomerId', card.customerId));
  }

  return items.filter((x): x is PaymentInfoItem => x !== null);
}

function mapTokenResponse(res: TokenCommandResponse): SavedCard {
  const arr = res.TokenInfoArray ?? [];
  const findAttr = (code: number) => arr.find((i) => i.AttributeCode === code)?.AttributeValue;

  return {
    tokenId: findAttr(1) ?? '',
    maskedCard: findAttr(12),
    last4: findAttr(11),
    paymentSystem: findAttr(2),
    fiCode: findAttr(9),
    fiName: findAttr(10),
    brandImageUrl: findAttr(13),
    requiresOneTimePassword: !!findAttr(25),
    lifetimeSecs: res.LifetimeSecs,
  };
}

export class TokensModule {
  private readonly config: EcollectConfig;
  private readonly http: HttpClient;
  private readonly session: SessionModule;

  constructor(config: EcollectConfig, http: HttpClient, session: SessionModule) {
    this.config = config;
    this.http = http;
    this.session = session;
  }

  private get baseUrl(): string {
    return BASE_URLS[this.config.environment];
  }

  private async _command(command: TokenCommand, tokenInfoArray: PaymentInfoItem[]): Promise<SavedCard> {
    const url = `${this.baseUrl}/tokenCommand`;
    const sessionToken = await this.session.getActive();

    const body: TokenCommandRequest = {
      EntityCode: this.config.etyCode,
      SessionToken: sessionToken,
      Command: command,
      TokenInfoArray: tokenInfoArray,
    };

    const res = await this.http.post<TokenCommandRequest, TokenCommandResponse>(url, body);

    if (res.ReturnCode === 'FAIL_APIEXPIREDSESSION') {
      this.session.invalidate();
      body.SessionToken = await this.session.getActive();
      const retried = await this.http.post<TokenCommandRequest, TokenCommandResponse>(url, body);
      if (retried.ReturnCode !== 'SUCCESS' && retried.ReturnCode !== 'SUCCESS_ALREADY_CREATED') {
        throw mapReturnCodeToError(retried.ReturnCode);
      }
      return mapTokenResponse(retried);
    }

    if (res.ReturnCode !== 'SUCCESS' && res.ReturnCode !== 'SUCCESS_ALREADY_CREATED') {
      throw mapReturnCodeToError(res.ReturnCode);
    }

    return mapTokenResponse(res);
  }

  /**
   * Save a card token persistently (SAVE command).
   */
  async save(cardData: CardData): Promise<SavedCard> {
    validateCardNumber(cardData.cardNumber);
    validateExpirationDate(cardData.expirationDate);
    return this._command('SAVE', buildCardTokenInfoArray(cardData));
  }

  /**
   * Get a temporary token without saving (GET command).
   */
  async get(cardData: CardData): Promise<SavedCard> {
    validateCardNumber(cardData.cardNumber);
    validateExpirationDate(cardData.expirationDate);
    return this._command('GET', buildCardTokenInfoArray(cardData));
  }

  /**
   * Get a hold token for pre-authorization (HOLD command).
   */
  async hold(cardData: CardData): Promise<SavedCard> {
    validateCardNumber(cardData.cardNumber);
    validateExpirationDate(cardData.expirationDate);
    return this._command('HOLD', buildCardTokenInfoArray(cardData));
  }

  /**
   * List saved cards for a user (queryToken).
   */
  async list(email: string, cardHolderId: string): Promise<SavedCard[]> {
    const url = `${this.baseUrl}/queryToken`;
    const sessionToken = await this.session.getActive();

    const tokenInfoArray: PaymentInfoItem[] = [
      { AttributeCode: 6, AttributeDesc: 'Usermail', AttributeValue: email },
      { AttributeCode: 19, AttributeDesc: 'CardHolderId', AttributeValue: cardHolderId },
    ];

    const body: QueryTokenRequest = {
      EntityCode: this.config.etyCode,
      SessionToken: sessionToken,
      TokenInfoArray: tokenInfoArray,
    };

    const res = await this.http.post<QueryTokenRequest, QueryTokenResponse>(url, body);

    if (res.ReturnCode === 'NO_RECORDS') return [];
    if (res.ReturnCode !== 'SUCCESS') {
      throw mapReturnCodeToError(res.ReturnCode);
    }

    return (res.TokenArray ?? []).map((record) => {
      const arr = record.TokenInfoArray ?? [];
      const findAttr = (code: number) => arr.find((i) => i.AttributeCode === code)?.AttributeValue;
      return {
        tokenId: findAttr(1) ?? '',
        maskedCard: findAttr(12),
        last4: findAttr(11),
        bin4: findAttr(30),
        paymentSystem: findAttr(2),
        fiCode: findAttr(9),
        fiName: findAttr(10),
        brandImageUrl: findAttr(13),
        email: findAttr(6),
        customerId: findAttr(100),
        tokenStatus: record.TokenStatus as SavedCard['tokenStatus'],
        lifetimeSecs: record.LifetimeSecs,
        requiresOneTimePassword: !!findAttr(25),
      };
    });
  }

  /**
   * Delete a saved card (REMOVE command).
   */
  async delete(
    tokenId: string,
    email: string,
    cardHolderId: string,
  ): Promise<void> {
    await this._command('REMOVE', [
      { AttributeCode: 1, AttributeDesc: 'TokenId', AttributeValue: tokenId },
      { AttributeCode: 6, AttributeDesc: 'Usermail', AttributeValue: email },
      { AttributeCode: 19, AttributeDesc: 'CardHolderId', AttributeValue: cardHolderId },
    ]);
  }

  /**
   * Update the expiration date of a saved card (UPDATE command).
   */
  async update(
    tokenId: string,
    newExpiry: string,
    cardHolderId?: string,
  ): Promise<SavedCard> {
    validateExpirationDate(newExpiry);
    const items: PaymentInfoItem[] = [
      { AttributeCode: 1, AttributeDesc: 'TokenId', AttributeValue: tokenId },
      { AttributeCode: 4, AttributeDesc: 'ExpirationDate', AttributeValue: newExpiry },
    ];
    if (cardHolderId) {
      items.push({ AttributeCode: 19, AttributeDesc: 'CardHolderId', AttributeValue: cardHolderId });
    }
    return this._command('UPDATE', items);
  }
}

File: packages/python/ecollect/modules/tokens.py
"""Token management module (tokenCommand + queryToken)."""
import logging
from typing import Any, Dict, List, Optional

from ecollect.config import EcollectConfig
from ecollect.exceptions import raise_for_return_code
from ecollect.types import SavedCard
from ecollect.utils.http import AsyncHttpClient, run_sync

logger = logging.getLogger(__name__)


def _info_item(code: int, desc: str, value: str) -> Dict[str, Any]:
    return {"AttributeCode": code, "AttributeDesc": desc, "AttributeValue": str(value)}


def _parse_saved_card(token_info_array: List[Dict[str, Any]]) -> SavedCard:
    """Parse a TokenInfoArray list into a SavedCard object."""
    attrs: Dict[int, str] = {
        item["AttributeCode"]: item.get("AttributeValue", "")
        for item in token_info_array
    }
    return SavedCard(
        token_id=attrs.get(1, ""),
        masked_card=attrs.get(12),
        last4=attrs.get(11),
        fi_code=attrs.get(9),
        fi_name=attrs.get(10),
        payment_system=attrs.get(2),
        brand_image_url=attrs.get(13),
        customer_id=attrs.get(36),  # CustomerId attr code
    )


class TokensModule:
    """Handles tokenisation commands and saved card queries."""

    def __init__(
        self,
        config: EcollectConfig,
        http: AsyncHttpClient,
        session: "SessionModule",  # type: ignore[name-defined]
    ) -> None:
        self._config = config
        self._http = http
        self._session = session

    async def _post(self, endpoint: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        url = self._config.endpoint_url(endpoint)
        data = await self._http.post_with_retry(url, payload)
        return_code = data.get("ReturnCode", "")
        if return_code == "FAIL_APIEXPIREDSESSION":
            self._session.invalidate()
            payload["SessionToken"] = await self._session.get_active()
            data = await self._http.post_with_retry(url, payload)
        raise_for_return_code(data.get("ReturnCode", ""), str(data))
        return data

    async def _token_command(self, command: str, token_info: List[Dict[str, Any]]) -> Dict[str, Any]:
        session_token = await self._session.get_active()
        payload = {
            "EntityCode": self._config.ety_code,
            "SessionToken": session_token,
            "Command": command,
            "TokenInfoArray": token_info,
        }
        return await self._post("tokenCommand", payload)

    # ------------------------------------------------------------------
    # Async API
    # ------------------------------------------------------------------

    async def save(self, token_info: List[Dict[str, Any]]) -> Dict[str, Any]:
        """SAVE command — tokenize a card for future use."""
        return await self._token_command("SAVE", token_info)

    async def get_temporary(self, token_info: List[Dict[str, Any]]) -> Dict[str, Any]:
        """GET command — obtain a temporary token (single-use, no persistence)."""
        return await self._token_command("GET", token_info)

    async def hold(self, token_info: List[Dict[str, Any]]) -> Dict[str, Any]:
        """HOLD command — obtain a reserved token for pre-authorization."""
        return await self._token_command("HOLD", token_info)

    async def list_saved(self, usermail: str, card_holder_id: str) -> List[SavedCard]:
        """Query all saved cards for a user via queryToken."""
        session_token = await self._session.get_active()
        payload = {
            "EntityCode": self._config.ety_code,
            "SessionToken": session_token,
            "TokenInfoArray": [
                _info_item(6, "Usermail", usermail),
                _info_item(19, "CardHolderId", card_holder_id),
            ],
        }
        data = await self._post("queryToken", payload)
        token_array: List[Dict[str, Any]] = data.get("TokenArray", [])
        return [_parse_saved_card(t.get("TokenInfoArray", [])) for t in token_array]

    async def delete(self, token_id: str, card_holder_id: Optional[str] = None) -> Dict[str, Any]:
        """REMOVE command — delete a saved card token."""
        token_info = [_info_item(1, "TokenId", token_id)]
        if card_holder_id:
            token_info.append(_info_item(19, "CardHolderId", card_holder_id))
        return await self._token_command("REMOVE", token_info)

    async def update(self, token_id: str, expiration_date: str) -> Dict[str, Any]:
        """UPDATE command — update the expiration date of a saved card."""
        token_info = [
            _info_item(1, "TokenId", token_id),
            _info_item(4, "ExpirationDate", expiration_date),
        ]
        return await self._token_command("UPDATE", token_info)

    # ------------------------------------------------------------------
    # Sync wrappers
    # ------------------------------------------------------------------

    def save_sync(self, token_info: List[Dict[str, Any]]) -> Dict[str, Any]:
        return run_sync(self.save(token_info))

    def get_temporary_sync(self, token_info: List[Dict[str, Any]]) -> Dict[str, Any]:
        return run_sync(self.get_temporary(token_info))

    def hold_sync(self, token_info: List[Dict[str, Any]]) -> Dict[str, Any]:
        return run_sync(self.hold(token_info))

    def list_saved_sync(self, usermail: str, card_holder_id: str) -> List[SavedCard]:
        return run_sync(self.list_saved(usermail, card_holder_id))

    def delete_sync(self, token_id: str, card_holder_id: Optional[str] = None) -> Dict[str, Any]:
        return run_sync(self.delete(token_id, card_holder_id))

    def update_sync(self, token_id: str, expiration_date: str) -> Dict[str, Any]:
        return run_sync(self.update(token_id, expiration_date))
